### PR TITLE
[JENKINS-27120] Adding Workflow support for JaCoCo publisher

### PR DESCRIPTION
**Pull request by centic9**
Continued from PR #63 without all the whitespace changes...

### CODE REVIEW

**Comment by andresrc** [comment]
:-1: not keeping the original authorship. Not polite and discourages contribution.

**Comment by lordofthejars** [comment]
oh please, I would like to have this issue fixed and released by Thursday, anyone with permissions think it is going to be possible?

**Comment by portenez** [comment]
Is there any chance this could be pulled/merged in the near future?

@andresrc, I think @lordofthejars is more concerned with getting this released than with the authorship, and I could also use this fix. Also, there's the link to the original PR, which will show the original authorship. (@lordofthejars)

In any case, thanks all.

**Comment by lordofthejars** [comment]
Exactly, I don t care the authorship I want that people can improve how
they are working.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 1]
```diff
@@ -359,4 +355,14 @@ public final PrintStream getLogger() {
 	    // does not run the construct and thus leaves the transient variables empty
 	    return System.out;
 	}
+
+	@Override
+	public void onAttached(Run<?, ?> run) {
+		this.owner = run;
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Note that this means you could delete the `owner` parameter from the constructor (assuming it is not used within the dynamic scope of the constructor).

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 2]
```diff
@@ -302,7 +400,18 @@ protected static String resolveFilePaths(AbstractBuild<?, ?> build, BuildListene
     @SuppressWarnings("resource")
     @Override
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: You need not retain this overload—it is implemented in the superclass to delegate to the new one.

Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: (So by deleting it you can inline `publishReports` again.)

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 3]
```diff
@@ -471,6 +584,39 @@ public Publisher newInstance(StaplerRequest req, JSONObject json) throws FormExc
...
+            String[] dirs = ds.getIncludedDirectories();
+    
+            for (String dir : dirs) {
+                localDirectoryPaths.add(base.child(dir));
+            }
+            FilePath[] lfp = {};//trick to have an empty array as a parameter, so the returned array will contain the elements
+            return localDirectoryPaths.toArray(lfp);
+        }
+        
+        public void checkRoles(RoleChecker checker) throws SecurityException {
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: :bug: extend `MasterToSlaveFileCallable` instead.

**Comment by jglick** [comment]
Other than the incorrect use of `checkRoles`, looks basically good. Seems it need a remerge with `master` though.

**Comment by lordofthejars** [comment]
@centic9 are you going to continue with this issue or you prefer I get it again and apply the chagnes suggested by @jglick

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 1 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: @jglick so I can remove the first parameter of this constructor: `public JacocoBuildAction(Run<?,?> owner, Map<CoverageElement.Type, Coverage> ratios, ...)` without worrying about any side effect. Other plugins currently are not using JaCoCo API.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 2 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: @jglick yes then it has no sense at all

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 3 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: @jglick ok, pretty fast to do this :)

**Comment by centic9** [comment]
Just go ahead please, I actually lack the insight into the Jenkins internals necessary here anyway, just let me know when it is ready to merge

**Comment by lordofthejars** [comment]
@centic9 @jglick The changes you mention has been merged to this PR. Do we need bees or are not necessary since the PR was not originated by me. Of course a review about I have done the cahnges correctly would be much appreciated.

File: pom.xml [hunk 4]
```diff
@@ -92,6 +92,7 @@ THE SOFTWARE.
       <project.build.outputEncoding>UTF-8</project.build.outputEncoding>
       <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
       <jacoco.version>0.7.5.201505241946</jacoco.version>
+        <workflow.version>1.12</workflow.version>
```
Comment (jglick) [inline pom.xml]: Why is this here? Seems unused. (Of course you would use it if you wanted to add automated tests.)

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 1 above)
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Right. Note that the `load` method will return an action whose `getBuild()` would at that time return null, but once you call `someBuild.add(theNewAction)` it gets initialized.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 5]
```diff
@@ -294,107 +393,109 @@ protected static String resolveFilePaths(AbstractBuild<?, ?> build, BuildListene
...
-		if (build.getResult() == Result.FAILURE || build.getResult() == Result.ABORTED) {
-			return true;
-		}
-		
-		
-		logger.println("[JaCoCo plugin] Collecting JaCoCo coverage data...");
-		
-		
-		EnvVars env = build.getEnvironment(listener);
-        env.overrideAll(build.getBuildVariables());
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: This seems to have been deleted without replacement for freestyle builds.

Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: I think you wanted above:

```java
Map<String, String> envs = run instanceof AbstractBuild ? ((AbstractBuild) run).getBuildVariables() : Collections.<String, String>emptyMap();
```

or similar.

Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: totally agree, sorry @jglick this my first plugin collaboration :laughing:

**Comment by lordofthejars** [comment]
@jglick ok I think that now it is ready to be merged into master. If you agree let's put a bee and I talk with James for the same.

**Comment by lordofthejars** [comment]
@reviewbybees

**Comment by jmvanryn** [comment]
Is there any update on this?

**Comment by lordofthejars** [comment]
For me it is ok, but after pushing hard for trying to release this and after the time I decided to give it up.

**Comment by isubuz** [comment]
Are there any plans to merge this soon?

**Comment by lordofthejars** [comment]
I think that answer should be done or by @centic9 or by @jglick and merge it. I cannot do anything else apart from sending the original PR.

**Comment by mscharp** [comment]
@lordofthejars it seems you and @jglick are both Bees employees... can you email/call him if you still need this PR reviewed?  Considering its been a month since the last update, maybe this PR just got lost in the ether?

**Comment by centic9** [comment]
I will allow for a few more days to let others do reviews and then merge the piece. It looks basically ok to me, I only lack any knowledge about the Workflow based stuff to decide if that is done correctly or not, but at least it passes tests and still works for normal jobs then.

**Comment by jglick** [comment]
This is on my shortlist to review, just have not gotten to it yet, maybe today…

File: pom.xml [hunk 6]
```diff
@@ -28,8 +28,8 @@ THE SOFTWARE.
    <parent>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>plugin</artifactId>
-      <version>1.424.6</version>
-      <relativePath />
+      <version>1.609.3</version>
```
Comment (jglick) [inline pom.xml]: BTW consider using the 2.x parent POM at some point. A lot more pleasant to work with. (The baseline would in that case be defined using the `jenkins.version` property.)

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 7]
```diff
@@ -33,9 +36,9 @@
  * @author Jonathan Fuerth
  * @author Ognjen Bubalo
  */
-public final class JacocoBuildAction extends CoverageObject<JacocoBuildAction> implements HealthReportingAction, StaplerProxy, Serializable {
+public final class JacocoBuildAction extends CoverageObject<JacocoBuildAction> implements HealthReportingAction, StaplerProxy, Serializable, RunAction2 {
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Good, the non-`transient` field could actually cause (non-Pipeline-related) problems as of 1.483.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 8]
```diff
@@ -202,12 +196,12 @@ public Object getTarget() {
 	}
 
 	@Override
-	public AbstractBuild<?,?> getBuild() {
+	public Run<?,?> getBuild() {
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: This and similar changes are not binary compatible, but I could not find [anything in @jenkinsci referring to code in this plugin](https://github.com/search?utf8=%E2%9C%93&q=user%3Ajenkinsci+-repo%3Ajenkinsci%2Fjacoco-plugin+jacoco&type=Code&ref=searchresults) so probably it is fine.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 9]
```diff
@@ -55,33 +68,54 @@
...
+    private String maximumLineCoverage;
+    private String maximumMethodCoverage;
+    private String maximumClassCoverage;
+    private boolean changeBuildStatus;
     
 	private static final String DIR_SEP = "\\s*,\\s*";
 
+    @DataBoundConstructor
+    public JacocoPublisher() {
+        this.execPattern = "**/**.exec";
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: BTW if you want to avoid duplicating defaults with `config.jelly`, see [this trick](https://github.com/jenkinsci/workflow-plugin/blob/d9234cc89d6bf7245d0d1f71e96faf8d2ea99ee4/COMPATIBILITY.md#handling-default-values).

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 10]
```diff
@@ -55,33 +68,54 @@
...
+        this.classPattern = "**/classes";
+        this.sourcePattern = "**/src/main/java";
+        this.inclusionPattern = "";
+        this.exclusionPattern = "";
+        this.minimumInstructionCoverage = "0";
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Would be a good opportunity to change the type of these parameters to (I guess) `int`, which is more comfortable for use from scripted clients. Can retain `String` in the serial (XStream) form for compatibility.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 11]
```diff
@@ -240,14 +274,103 @@ public boolean isChangeBuildStatus() {
...
+        this.maximumMethodCoverage = checkThresholdInput(maximumMethodCoverage);
+    }
+
+    @DataBoundSetter
+    public void setMaximumClassCoverage(String maximumClassCoverage) {
+        this.maximumClassCoverage = checkThresholdInput(maximumClassCoverage);
+    }
+
+    @DataBoundSetter
+    public void setChangeBuildStatus(boolean changeBuildStatus) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Not really part of this PR, but BTW the `checked="${instance.selected}"` in `config.jelly` seems to be a typo, and could simply be deleted anyway.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 12]
```diff
@@ -259,142 +382,121 @@ protected static String resolveFilePaths(AbstractBuild<?, ?> build, BuildListene
...
+
+        if (run.getResult() == Result.FAILURE || run.getResult() == Result.ABORTED) {
+            return;
+        }
+
+
+        logger.println("[JaCoCo plugin] Collecting JaCoCo coverage data...");
+
+
+        EnvVars env = run.getEnvironment(taskListener);
```
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Note that due to [JENKINS-29144](https://issues.jenkins-ci.org/browse/JENKINS-29144) this will not handle Pipeline `withEnv`.

Anyway `env` seems to be unused. I think you meant to pass it to `resolveFilePaths`.

Anyway, for Pipeline, we do not _want_ anything to be substituted. `execPattern` should be taken as a literal. So what you probably actually wanted was something like

```java
String _execPattern;
if (build instanceof AbstractBuild) {
    _execPattern = resolveFilePaths((AbstractBuild) build, listener, execPattern);
} else {
    _execPattern = execPattern;
}
…use as: filePath.list(_execPattern)
```

where you fix the helper method

```java
protected static String resolveFilePaths(AbstractBuild<?, ?> build, TaskListener listener, String input) {
    EnvVars env = build.getEnvironment(listener);
    env.overrideAll(build.getBuildVariables());
    // …as before
```

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 10 above)
Comment (jglick) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Actual you could even change the field type to `int` without causing problems for XStream deserialization, I think. Would avoid the need to catch `NumberFormatException` when constructing the `JacocoHealthReportThresholds`.

File: src/test/java/hudson/plugins/jacoco/report/AbstractReportTest.java [hunk 13]
```diff
@@ -26,41 +27,38 @@ public void test() throws Exception {
...
         report.setParent(new ClassReport());
         report.getParent().setParent(new PackageReport());
-        
-        JacocoBuildAction action = new JacocoBuildAction(null, null, null, null, new BuildListener() {
-            
-            public void hyperlink(String url, String text) throws IOException {
-            }
-            
+
+        TaskListener taskListener = new TaskListener() {
```
Comment (jglick) [inline src/test/java/hudson/plugins/jacoco/report/AbstractReportTest.java]: `StreamTaskListener.fromStdout()` would be easier.

File: src/test/java/hudson/plugins/jacoco/report/CoverageReportTest.java [hunk 14]
```diff
@@ -44,7 +45,7 @@ public void testThresholds() throws Exception {
         report.setThresholds(new JacocoHealthReportThresholds());
     }
 
-    private JacocoBuildAction action = new JacocoBuildAction(null, null, null, null, new BuildListener() {
+    private JacocoBuildAction action = new JacocoBuildAction(null, null, new TaskListener() {
```
Comment (jglick) [inline src/test/java/hudson/plugins/jacoco/report/CoverageReportTest.java]: ditto

File: src/test/java/hudson/plugins/jacococoveragecolumn/JaCoCoColumnTest.java [hunk 15]
```diff
@@ -85,7 +86,7 @@ public void testGetPercentWithBuildAndAction() {
 			public Run<?,?> getLastSuccessfulBuild() {
 				try {
 				    Run<?,?> newBuild = newBuild();
-					newBuild.getActions().add(new JacocoBuildAction(null, null, null, null, new BuildListener() {
+					newBuild.getActions().add(new JacocoBuildAction(null, null, new TaskListener() {
```
Comment (jglick) [inline src/test/java/hudson/plugins/jacococoveragecolumn/JaCoCoColumnTest.java]: ditto

**Comment by isubuz** [comment]
Sorry for nudging, but @centic9 do you plan to address comments anytime soon? Basically my team is trying to migrate to the workflow plugin completely and this is the plugin which is not yet compatible, hence the urgency.

**Comment by centic9** [comment]
Sorry, but I am only a facilitator here as the original changes came from @lordofthejars , hopefully he can take a look soon.

**Comment by isubuz** [comment]
@lordofthejars ?

**Comment by lordofthejars** [comment]
Currently we are putting the efforts on Jenkins 2. So probably it will take some time.

**Comment by lordofthejars** [comment]
@centic9 I will need to work on your branch, so the PR will come to your branch and repo and when you accept it then it will be reflected here.

File: pom.xml (hunk 6 above)
Comment (lordofthejars) [inline pom.xml]: For now I will leave as is and if in the future we need to move forward, let's open an issue. Changing parent pom could generate any incompatibility? For what I see is that it is only going to work in Jenkins 2?

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 7 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: We are using 1.609.3 so is there any problem or it is ok as I have coded?

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 8 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Exactly I think it is ok in this way.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 1 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: Done

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 10 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: I agree with you but since originally the project was using an String , I think that this change should be done in another PR for refactoring to not mix everything, and maintain this PR as the minimal changes to make this work for Pipeline.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 11 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Let's open a new PR to not mix things in this PR and maintain it as minimal as possible, just having minimal changes to make it work in pipeline

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 5 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Done

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 12 above)
Comment (lordofthejars) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Done

File: src/test/java/hudson/plugins/jacoco/report/AbstractReportTest.java (hunk 13 above)
Comment (lordofthejars) [inline src/test/java/hudson/plugins/jacoco/report/AbstractReportTest.java]: Done

File: src/test/java/hudson/plugins/jacoco/report/CoverageReportTest.java (hunk 14 above)
Comment (lordofthejars) [inline src/test/java/hudson/plugins/jacoco/report/CoverageReportTest.java]: Done

**Comment by centic9** [comment]
Sure, just let me know when you think it is ready for merge so we can finally release this.

**Comment by faheem-cliqz** [comment]
Hi guys an ETA on this... we are also migrating to pipeline plugin, and it seems this essential plugin is missing :(

**Comment by lordofthejars** [comment]
@jglick must review the last changes.

**Comment by bengineer-** [comment]
I'm also moving to the Pipelne plugin and this PR would be a big help.i

**Comment by centic9** [comment]
In order to allow others to give feedback more easily I have published a pre-release with these changes at https://github.com/jenkinsci/jacoco-plugin/releases/tag/2.1.0-beta1, please give it a try by manually installing the .hpi file in the Jenkins installation via "manage jenkins - plugins - advanced options - upload plugin" and report here if it works for you or not.

**Comment by centic9** [comment]
BTW, a local build fails for me with the following on 1.638, seems the onAttached() is not called there? The changes set the required jenkins version to 1.609.3.

ERROR: Build step failed with exception
java.lang.NullPointerException
	at hudson.plugins.jacoco.JacocoBuildAction.getJacocoReport(JacocoBuildAction.java:204)
	at hudson.plugins.jacoco.JacocoBuildAction.getResult(JacocoBuildAction.java:219)
	at hudson.plugins.jacoco.JacocoPublisher.perform(JacocoPublisher.java:490)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:76)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)

**Comment by leifbladt** [comment]
I can confirm the mentioned `NullPointerException` with Jenkins 2.2

**Comment by mheinzerling** [comment]
Created  PR#73 to fix the NPE.

**Comment by centic9** [comment]
After the NPE was fixed via #73 I have now merged these changes onto trunk. Please test it some more by running "mvn package" and then importing the resulting target/jacoco.hpi manually in Jenkins -> Manage -> Plugins -> Advanced.

**Comment by anyone-can-test** [comment]
I tested the latest jacoco plugin (mvn package in my local pc) and the pipeline functionality.
The coverage result was good.
But, there was no Code Coverage Trend in the main screen of a job.

**Comment by mheinzerling** [comment]
Might be reletad to JENKINS-31202.

**Comment by mscharp** [comment]
Could someone post, or link to a post that shows, what the syntax would be in order to use the JacocoPublisher?

**Comment by mheinzerling** [comment]
Just run the current master or even mheinzerling_2.1.0 with a current jenkins and you will get the pipeline configuration util for free. You can configure your step with the usual graphical interface.

**Comment by mpchlets** [comment]
@mheinzerling - when using pipeline from a Jenkinsfile, you cannot use the graphical interface to configure.

**Comment by mheinzerling** [comment]
@mpchlets, more details please. Pipeline and GUI are totally unrelated.

**Comment by jreimone** [comment]
How to use the JacocoPublisher in a pipeline script? I cannot find any syntactical descriptions. I just tried to add `step([$class: 'JaCoCoPublisher'])` into my script but I get `java.lang.UnsupportedOperationException: no known implementation of interface jenkins.tasks.SimpleBuildStep is named JaCoCoPublisher` with JaCoCo plubin v.2.0.1

**Comment by thatsIch** [comment]
@jreimone This is not supported in 2.0.1. Update at least to 2.1.0-beta

**Comment by jreimone** [comment]
@thatsIch, thank you very much. With 2.1.0-beta release I got an NPE but it worked successfully with the latest master I Built locally. Is there a possibility to display the trend of the coverage in the job's overview, as it is also done by the JUnit test results?

**Comment by thatsIch** [comment]
You can use the Jacocopublisher to do that, this allows to display it in each build (in #1, #2 etc on the left side). Sadly the overview on the dashboard is not working since some integration is not working.

**Comment by tadgh** [comment]
Would really like the ability to have it in dashboard, but in build is good enough for now! I'm in the same place as Jreimone

**Comment by thatsIch** [comment]
You can also use the html publisher to publish the `index.html` of JaCoCo and give you a quick link to the latest one. Every solution sadly sub-optimal, but better than every dev stalling his IDE waiting for his coverage

**Comment by vladdu** [comment]
I just tried `step([$class: 'JaCoCoPublisher'])` with 2.1.0 and I get "no known implementation of interface jenkins.tasks.SimpleBuildStep is named JaCoCoPublisher"... Should it work?

**Comment by thatsIch** [comment]
It is called `JacocoPublisher`

**Comment by vladdu** [comment]
Oh, silly me, I just copied from a comment above and didn't check... Thanks!

**Comment by thatsIch** [comment]
Always look into the source :P it is actually easier than to rely on hear-say. That is how I know about the class name.

You are welcome :)
